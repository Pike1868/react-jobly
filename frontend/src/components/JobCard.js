import React, { useState, useEffect } from "react";
import {
  Button,
  Card,
  CardActions,
  CardContent,
  Grid,
  Typography,
} from "@mui/material";
import { Link } from "react-router-dom";
import JoblyApi from "../api/JoblyApi";
import { useUserContext } from "../context/UserContext";

export default function JobCard({ job }) {
  const { userDetails, setUserDetails } = useUserContext();
  const { id, title, salary, equity } = job;

  // Check if user has already applied for this job
  const [hasApplied, setHasApplied] = useState(
    userDetails.applications.includes(id)
  );

  useEffect(() => {
    // Update the state when userDetails.applications changes
    setHasApplied(userDetails.applications.includes(id));
  }, [userDetails.applications, id]);

  const handleApplyClick = async () => {
    try {
      await JoblyApi.applyToJob(userDetails.username, id);
      setUserDetails({
        ...userDetails,
        applications: [...userDetails.applications, id],
      });
    } catch (err) {
      console.error(err);
    }
  };
  return (
    <Grid item xs={12} sm={6} md={4}>
      <Card
        sx={{
          height: "100%",
          display: "flex",
          flexDirection: "column",
        }}
      >
        <CardContent sx={{ flexGrow: 1 }}>
          <Typography gutterBottom variant="h5" component="h2">
            {title}
          </Typography>
          <Typography align="left">Salary: {salary}</Typography>
          <Typography variant="body2" align="left" sx={{ padding: "16px 0 0" }}>
            Equity: {equity}
          </Typography>
        </CardContent>

        <CardActions>
          <Link to="#" style={{ textDecoration: "none" }}>
            <Button
              size="small"
              onClick={() => {
                handleApplyClick();
              }}
              disabled={hasApplied}
            >
              {hasApplied ? "Applied" : "Apply"}
            </Button>
          </Link>
        </CardActions>
      </Card>
    </Grid>
  );
}
