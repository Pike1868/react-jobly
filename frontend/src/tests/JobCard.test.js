import React from "react";
import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import JobCard from "../components/JobCard";
import { mockUserDetails } from "./commonSetup"; 

jest.mock("../context/UserContext", () => ({
  useUserContext: () => mockUserDetails, 
}));

describe("JobCard Component", () => {
  const mockJob = {
    id: 1,
    title: "Software Engineer",
    salary: 120000,
    equity: "0.05",
  };

  it("renders without crashing", () => {
    render(
      <MemoryRouter>
        <JobCard job={mockJob} />
      </MemoryRouter>
    );
  });

  it("matches snapshot", () => {
    const { asFragment } = render(
      <MemoryRouter>
        <JobCard job={mockJob} />
      </MemoryRouter>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
