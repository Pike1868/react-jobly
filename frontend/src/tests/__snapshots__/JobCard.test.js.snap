// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JobCard Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-4 css-13i4rnv-MuiGrid-root"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-16vq84g-MuiPaper-root-MuiCard-root"
    >
      <div
        class="MuiCardContent-root css-2r3irp-MuiCardContent-root"
      >
        <h2
          class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-h93ljk-MuiTypography-root"
        >
          Software Engineer
        </h2>
        <p
          class="MuiTypography-root MuiTypography-body1 MuiTypography-alignLeft css-zpphrz-MuiTypography-root"
        >
          Salary: 120000
        </p>
        <p
          class="MuiTypography-root MuiTypography-body2 MuiTypography-alignLeft css-nitxol-MuiTypography-root"
        >
          Equity: 0.05
        </p>
      </div>
      <div
        class="MuiCardActions-root MuiCardActions-spacing css-dnrpxu-MuiCardActions-root"
      >
        <a
          href="/"
          style="text-decoration: none;"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall Mui-disabled MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall css-1knaqv7-MuiButtonBase-root-MuiButton-root"
            disabled=""
            tabindex="-1"
            type="button"
          >
            Applied
          </button>
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
